### PR TITLE
Add internal AST testing support

### PR DESCRIPTION
**Pull request by leonard84**

### CODE REVIEW

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  def "astToSourceFeatureBody renders only methods and its annotation by default"() {
+    when:
+    def result = compiler.astToSourceFeatureBody('''
+    given:
+    def nothing = null
+    ''')
+
+    then:
+    result == '''\
+@org.spockframework.runtime.model.FeatureMetadata(name = 'a feature', ordinal = 0, line = 1, blocks = [org.spockframework.runtime.model.BlockKind.SETUP[]], parameterNames = [])
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:18]: The `blocks` value does not look right.
If you look at a decompiled class file of our tests, it looks like
```
blocks = {@BlockMetadata(
    kind = BlockKind.SETUP,
    texts = {"an a"}
)}
```

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    }
+    ''')
+
+    then:
+    result == '''\
+@org.spockframework.runtime.model.FieldMetadata(name = 'foo', ordinal = 0, line = 1, initializer = true)
+private java.lang.Object foo
+
+private java.lang.Object $spock_initializeFields() {
+    foo = 'bar\'
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:43]: Is this backslash significant?
It should be superfluous, shouldn't it?

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+}
+package apackage
+
+import groovy.transform.EqualsAndHashCode as EqualsAndHashCode
+
+@groovy.transform.EqualsAndHashCode
+public class apackage.Bar extends apackage.Foo {
+
+    private static final [I ARR
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:158]: Can we render this properly as `int[]`?
And where is the initializer?

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    private static final java.lang.String STR = 'str\'
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:159]: superfluous backslash

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    private static final char CHR = ' \'
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:160]: superfluous backslash

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+}
+package apackage
+
+import groovy.transform.EqualsAndHashCode as EqualsAndHashCode
+
+public class apackage.Ext<T extends java.io.Serializable> extends java.lang.Object {
+
+    public java.lang.Object foo(java.util.List<? super java.io.Serializable<T extends java.io.Serializable>> consumer) {
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:205]: shouldn't this be `List<? super T`>`?

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        order = 1
+    }
+
+    static {
+        st = 42
+    }
+
+    public void method() {
+        outer:
+        for (java.lang.Integer i = 0; i < INT ;( i )++) {
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:179]: Can we leave out the parens around the `i` if it is a simple variable expression?

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+            break outer
+        }
+        for (java.lang.String y : x ) {
+            this.print(y)
+        }
+    }
+
+    @java.lang.Override
+    public int compareTo(java.lang.Object o) {
+        return order <=> (( o ) as apackage.Bar).order
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:195]: Can we leave out the parens around the `i` if it is a simple variable expression?

**Review by Vampire — CHANGES_REQUESTED** [review]
I didn't do an in-depth review of the `AstNodeToSourceConverter`, just skimmed over it.
I guess most should be fine if taken from Groovy sources and I have seen you also fixed the array initializer issue, which is great, thanks. (Maybe you could PR it back to Groovy, I always miss it in the Groovy Console and stumbled due to it several times already)

But some points I've seen when skimming over it that you might consider:
- You wrote "made backwards compatible with groovy 2.4", do you mean 2.5? Or why did you made it 2.4 compatible? We only support 2.5 and 3.0 with Spock 2, or are there plans I missed? :-)
- I guess there was a `main` method you removed, but the class JavaDoc still says you can invoke it from the command line
- If you use a hard-coded Skriptname, you don't have to replace it later for comparison reasons, just always use `skript.groovy` and have that in the expected value
- There are error messages about "press Refresh" which don't make sense outside the context of the Groovy Console GUI application
- The class does too much, it is called `AstNodeToSourceConverter`, but that's not what it does. The name suggests it gets an ast node and converts that to source, but it gets a source string, compiles that to an AST and then converts that ast back to source. But I'm not suggesting to rename the class, but to split it, so that the compiling part can be reused for direct AST tests using Groovy object navigation. For some things we need the direct AST tests anyway as the source does not express all properties of the AST so there could be quite some cases where you check the decompiled AST against an expected value and also do some additional direct AST tests. So it would be nice if this could be designed so that you can do both, with only compiling the sources once and then check the decompiled script against an expected value and do some AST tests. Maybe there should actually be `ModuleNode compileXXXToAst` methods that give you the actual AST for fine-grained tests, one `astToSource` method that takes an AST and gives the source string and the existing `astToSourceXXX` methods will become `transpileXXX` and combine those other two methods, because that is what those methods do, transcompiling the source to another source.


I also skimmed over the `AstSpec` and annotated some things I've seen.
I guess there are probably some more places where the script renderer can need some improvements. :-)

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy [hunk 9]
```diff
@@ -75,35 +80,74 @@ class EmbeddedSpecCompiler {
...
+    doAstToSource(imports + source, showSet, phase)
+  }
+
+  private String doAstToSource(@Language('Groovy') String source, Set showSet, CompilePhase phase) {
+    loader.clearCache()
+    String result = new AstNodeToSourceConverter().compileToScript(source, phase.phaseNumber, showSet, loader)
+    // normalize result
+    result = FILENAME_PATTERN.matcher(result).replaceAll("XXXXX")
+    result = TRAILING_SPACES.matcher(result).replaceAll("")
+    result = StringGroovyMethods.stripIndent((CharSequence)result) // Java 15 introduces `stripIndent` with a different behavior, so use explicit method call
```
[the remark this answers is not in the document]
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:134]: `stripIndent((String)result)` is deprecated telling me to use `stripIndent((CharSequence)result)` so I did, to get rid of the warning

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy [hunk 10]
```diff
@@ -32,6 +35,8 @@ import org.opentest4j.MultipleFailuresError
  */
 @NotThreadSafe
 class EmbeddedSpecCompiler {
+  static final FILENAME_PATTERN = Pattern.compile(/(?<=filename = 'script)(\d+)(?=.groovy')/)
+  static final TRAILING_SPACES = Pattern.compile(/ +(?=\n)/)
```
[the remark this answers is not in the document]
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:39]: I'm pretty sure it is always `\n` as we write them via `printLineBreak()`, also I'm on Windows so it should have failed for me if that wasn't the case.
But is is actually not necessary anymore as I've improved `printLineBreak()` to remove trailing spaces.

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 3 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:158]: That is something still to be seen, if you look at the comments in `spock.util.AstNodeToScriptVisitor#visitField`, we might choose to change that behavior.

> // do not print initial expression, as this is executed as part of the constructor, unless on static constant

as for the type that is handled in `spock.util.AstNodeToScriptVisitor#visitType` this might require a switch statement for every primitive array type, or do you have a better suggestion

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 4 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:159]: Yeah, seems like intellij adds those when pasting

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 7 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:383]: It would be possible, but it would increase the complexity by introducing special case handling. 
IMHO the goal is to get a meaningful representation of the code, while keeping the renderer simple enough to maintain.

In any case I would see that as icing on top, and would defer it to a later change.

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy]: Ah, I see, sorry then :-)

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 3 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:362]: > // do not print initial expression, as this is executed as part of the constructor, unless on static constant

I cannot really follow the sense of that comment.
Why should where something is executed influence where it is in the script?
And even if it would make sense for the Groovy Console which I also dare to doubt, we want to check that the AST is like expected, so we should print it where it is in the AST.
Besides that it is simply swallowed completely, it is also not printed in the constructor but simply suppressed if it is not static final.
I would vote for always printing it in-place, that seems to be appropriate for our use-case imho.

> as for the type that is handled in spock.util.AstNodeToScriptVisitor#visitType this might require a switch statement for every primitive array type, or do you have a better suggestion

Maybe this is too naive and misses cases I'm not aware of, or it is simply that the class is 11 years old, but I'd say:
```groovy
  void visitType(ClassNode classNode) {
    print classNode.array ? classNode.toString(false) : classNode.name
    visitGenerics classNode?.genericsTypes
  }
```
The best would probably be we contribute those changes back upstream and see what the Groovy guys say and whether they see something else that is not working with the changes. :-)

Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:362]: Or probably
```groovy
  void visitType(ClassNode classNode) {
    if (classNode.array) {
      print classNode.toString(false)
    } else {
      print classNode.name
      visitGenerics classNode?.genericsTypes
    }
  }
```

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 6 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:467]: Maybe with something like this, but don't know about edge cases or similar, just from what I have seen in the debugger:
```groovy
if (it.lowerBound.genericsPlaceHolder) {
  print it.lowerBound.genericsTypes[0].name
} else {
  visitType it.lowerBound
}
```

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 7 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:383]: yeah, you are probably right. Should be enough if it were correct, does not have to be pretty.
A pitty that it is not really correct what is generated, the Groovy guys should put more love into this converter, they should know better what they are doing. :-D

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy [hunk 11]
```diff
@@ -75,35 +77,77 @@ class EmbeddedSpecCompiler {
...
   List<Class> compileWithImports(@Language('Groovy') String source) {
-    addPackageImport(Specification.package )
+    addPackageImport(Specification.package)
     // one-liner keeps line numbers intact
     doCompile "package apackage; $imports ${source.trim()}"
   }
 
   Class compileSpecBody(@Language(value = 'Groovy', prefix = 'class ASpec extends spock.lang.Specification { ', suffix = '\n }')
-                        String source) {
+                          String source) {
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:86]: Are these indentation changes intentional?

**Review by Vampire — CHANGES_REQUESTED** [review]

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy [hunk 12]
```diff
@@ -75,35 +77,77 @@ class EmbeddedSpecCompiler {
...
 
   Class compileFeatureBody(@Language(value = 'Groovy', prefix = "def 'a feature'() { ", suffix = '\n }')
-                           String source) {
+                             String source) {
     // one-liner keeps line numbers intact; newline safeguards against source ending in a line comment
     compileSpecBody "def 'a feature'() { ${source.trim() + '\n'} }"
   }
 
+  @Beta
+  AstResult compileToAstWithImports(@Language('Groovy') String source,
```
[the remark this answers is not in the document]
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:98]: I thought about separate methods, but think this overloads the API, if you care about CPU Cycles, you could just pass in an empty `Set` for `show` which basically disables the source output.

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy (hunk 11 above)
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:86]: Result of auto format file, so semi intentional ;)

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy [hunk 13]
```diff
@@ -0,0 +1,1237 @@
...
+      }
+    } else {
+      visitType expression?.leftExpression?.type
+      visitBinaryExpression expression // is a BinaryExpression
+    }
+  }
+
+  @Override
+  void visitGStringExpression(GStringExpression expression) {
+    // this loses enclosing ${}
```
[the remark this answers is not in the document]
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy:925]: Sounds good

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 6 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy]: Yeah generics handling needs to be improved

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 3 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:427]: I agree that we should print the field initializers.
The `visitType` needs to be improved anyway, since it does cause issues with generics.

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy [hunk 14]
```diff
@@ -0,0 +1,1237 @@
...
+   *    (optional) the classloader to use. If missing/null then the current is used.
+   *    This parameter enables things like ASTBrowser to invoke this with the correct classpath
+   * @param config
+   *    optional compiler configuration
+   * @returns the source code from the AST state and the captured ast nodes
+   */
+
+  AstResult compileScript(String script, int compilePhase, Set<Show> showSet, ClassLoader classLoader = null, CompilerConfiguration config = null) {
+
+    def writer = new StringBuilderWriter()
```
[the remark this answers is not in the document]
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy:52]: It wouldn't support rendering package and imports right? 
Any other things that wouldn't be supported?

We can decide, that we are not interested in those things anyway.
The final thing would need to support multiple classes., not just one.

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 1 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy]: yeah I fixed it now

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy (hunk 14 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy:52]: I've compared the two compilation methods and they are doing almost the same.
One difference is that you cannot give in an own class loader, but it uses a `new GroovyClassLoader()` but as far as I understood the own class loader in the embedded spec compiler is just to get `Class` instances out easily and you simply reused it because it was there anyway.
The other is that you don't get package statement and imports.

As the remaining things are identical I think there should be no further differences and I would agree that we don't necessarily need the package and imports ever in those tests.
And if we want them, we should anyway change the logic a bit, because currently the package and imports are repeated for every class node.

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy (hunk 12 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy]: Ah, good point, but let me try one last time, `transpileToAstSpecBody` sounds unnatural to me.
It sounds like "transpile to ast-spec-body" which makes me think what an "ast spec body" is.
If the "ast" has to be there, then I think "transpileSpecBodyToAst" would fit better.
While I still think that "ToAst" is not really fitting, transpile is "one source to other source" and the AST here is just an intermediate result that we also return together with the transpile result for direct AST checks.
I would prefer we have `transpileSpecBody` which returns a `TranspileResult` with result source and AST.

But if you want to keep it like it is now, I try to not mention it again. :-D

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy (hunk 11 above)
Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:85]: Really?
I thought about it, but couldn't come up with the settings that would produce this.
I know indent length, continuation indent length and aligned with the other parameters.
Which settings combination produces it like that?

Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:85]: https://github.com/leonard84/spock/blob/formatter-issue-555/config/idea/Spock.xml

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 1 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy]: Isn't the `org.spockframework.runtime.model.BlockKind.SETUP[]` in the beginning still wrong?
I think I would have expected
```
blocks = [@org.spockframework.runtime.model.BlockMetadata(kind = org.spockframework.runtime.model.BlockKind.SETUP, texts = [])]
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy [hunk 15]
```diff
@@ -0,0 +1,1237 @@
...
+    def scriptName = 'script.groovy'
+    GroovyCodeSource codeSource = new GroovyCodeSource(script, scriptName, '/groovy/script')
+    CompilationUnit cu = new CompilationUnit((CompilerConfiguration)(config ?: CompilerConfiguration.DEFAULT), (CodeSource)codeSource.codeSource, (GroovyClassLoader)classLoader)
+    def captureVisitor = new AstNoteCaptureVisitor()
+    cu.addPhaseOperation(captureVisitor, compilePhase)
+    cu.addPhaseOperation(new AstNodeToScriptVisitor(writer, showSet), compilePhase)
+    cu.addSource(codeSource.name, script)
+    try {
+      cu.compile(compilePhase)
+    } catch (CompilationFailedException cfe) {
```
[the remark this answers is not in the document]
Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy]: Hm, or maybe it is not bad, as we also intentionally produce compilation errors, so this might make it easier to test those.

File: spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy (hunk 11 above)
Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/EmbeddedSpecCompiler.groovy:85]: Ah, of course, this is aligend + continuation indent.
The `@Language` annoation is for that parameter, so the formatting makes sense of course.

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 3 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy]: I tested 
```
  void visitType(ClassNode classNode) {
      print classNode.toString(false)
  }
```

and it works well in most of the cases
```

public class apackage.another.Ext<T extends java.io.Serializable, V extends java.lang.Cloneable> extends java.lang.Object {

    private static final int[] ARR = (int[]) [1, 2, 3]
    private T[] arr
    private java.util.List <V> lst = []

    public V foo(java.util.List <? super T> consumer) {
    }

    @apackage.another.Ann
    public boolean saveCompare(X a, X b) {
    }
}
```

it does loose the information in `saveCompare` although the `<X extends java.io.Serializable & java.lang.Comparable<T>>` should actually be in front of the `boolean` here. Maybe you can think of a good heuristic on when to print it.

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 16]
```diff
@@ -0,0 +1,1241 @@
...
+      // GROOVY-5150: final constants may be initialized directly
+      print ' = '
+      if (ClassHelper.STRING_TYPE == type) {
+        print "'" + node.initialValueExpression.text.replace("'", "\\'") + "'"
+      } else if (ClassHelper.char_TYPE == type) {
+        print "'${node.initialValueExpression.text}'"
+      } else {
+        print node.initialValueExpression.text
+      }
+    }
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:568]:
```suggestion
    if (exp) {
      print ' = '
      exp.visit(this)
    }
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 17]
```diff
@@ -0,0 +1,1241 @@
...
+  void visitConstantExpression(ConstantExpression expression, boolean unwrapQuotes = false) {
+    if (expression instanceof AnnotationConstantExpression) {
+      visitAnnotationNode (expression.value as AnnotationNode)
+      return
+    }
+    if (expression.value instanceof String && !unwrapQuotes) {
+      // string reverse escaping is very naive
+      def escaped = ((String)expression.value).replaceAll('\n', '\\\\n').replaceAll("'", "\\\\'")
+      print "'$escaped'"
+    } else {
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:888]:
```suggestion
    if ((expression.value instanceof String || expression.value instanceof Character) && !unwrapQuotes) {
      // string reverse escaping is very naive
      def escaped = (expression.value as String).replaceAll('\n', '\\\\n').replaceAll("'", "\\\\'")
      print "'$escaped'"
    } else {
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 18]
```diff
@@ -0,0 +1,1241 @@
...
+  }
+
+  @Override
+  void visitCastExpression(CastExpression expression) {
+    print '(('
+    expression?.expression?.visit this
+    print ') as '
+    visitType(expression?.type)
+    print ')'
+
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:970]:
```suggestion
    if (expression.coerce) {
      print '(('
      expression?.expression?.visit this
      print ') as '
      visitType(expression?.type)
      print ')'
    } else {
      print '(('
      visitType(expression?.type)
      print ') '
      expression?.expression?.visit this
      print ')'
    }
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 19]
```diff
@@ -0,0 +1,1241 @@
...
+  void visitType(ClassNode classNode) {
+    def name = classNode.name
+    if (name =~ /^\[+L/ && name.endsWith(';')) {
+      int numDimensions = name.indexOf('L')
+      print "${classNode.name[(numDimensions + 1)..-2]}" + ('[]' * numDimensions)
+    } else {
+      print name
+    }
+    // see if we can improve rendering for method parameters
+    visitGenerics classNode?.genericsTypes
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:986]:
```suggestion
    if (classNode.array) {
      print classNode.toString(false)
    } else {
      print classNode.name
      // see if we can improve rendering for method parameters
      visitGenerics classNode?.genericsTypes
    }
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 20]
```diff
@@ -0,0 +1,1241 @@
...
+    if (node.name == '<init>') {
+      print "${classNameStack.peek()}("
+      visitParameters(node.parameters)
+      print ') {'
+      printLineBreak()
+    } else if (node.name == '<clinit>') {
+      print '{ ' // will already have 'static' from modifiers
+      printLineBreak()
+    } else {
+      visitType node.returnType
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:507]:
```suggestion
      if (node.genericsTypes) {
        visitGenerics node.genericsTypes
        print " "
      }
      visitType node.returnType
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 21]
```diff
@@ -0,0 +1,1241 @@
...
+      }
+      printLineBreak()
+      print '}'
+      printDoubleBreak()
+    }
+  }
+
+  private void visitGenerics(GenericsType[] generics) {
+
+    if (generics) {
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:428]:
```suggestion
    if (generics != null) {
```

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy (hunk 3 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy]: Part of the suggested changes, the method has generics information it is just not considered

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 22]
```diff
@@ -0,0 +1,1241 @@
...
+  }
+
+  private void visitExpressionsAndCommaSeparate(List<? super Expression> expressions) {
+    boolean first = true
+    expressions?.each {
+      if (!first) {
+        print ', '
+      }
+      first = false
+      ((ASTNode)it).visit this
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:1213]: I have no idea why the `AnnotationConstantExpression` does it, but it first visits the values of the members before it visits itself. That's why you got `org.spockframework.runtime.model.BlockKind.SETUP` followed by `[]` followed by the actual correct representation.
```suggestion
      if (it instanceof AnnotationConstantExpression) {
        visitConstantExpression(it)
      } else {
        (it as ASTNode).visit this
      }
```

Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:1213]: Reported it as https://issues.apache.org/jira/browse/GROOVY-9980

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 23]
```diff
@@ -0,0 +1,1241 @@
...
+  }
+
+
+  @Override
+  void visitClosureExpression(ClosureExpression expression) {
+    print '{ '
+    if (expression?.parameters) {
+      visitParameters(expression?.parameters)
+      print ' ->'
+    }
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:811]:
```suggestion
    }
    print ' ->'
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 24]
```diff
@@ -0,0 +1,1241 @@
...
+  void visitAttributeExpression(AttributeExpression attributeExpression) {
+    visitPropertyExpression attributeExpression
+  }
+
+  @Override
+  void visitFieldExpression(FieldExpression expression) {
+    print expression?.field?.name
+  }
+
+  void visitConstantExpression(ConstantExpression expression, boolean unwrapQuotes = false) {
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:879]:
```suggestion
  private static String escapeCharacters(String input) {
    input
      ?.replaceAll('\\\\', '\\\\\\\\')
      ?.replaceAll('\b', '\\\\b')
      ?.replaceAll('\f', '\\\\f')
      ?.replaceAll('\n', '\\\\n')
      ?.replaceAll('\r', '\\\\r')
      ?.replaceAll('\t', '\\\\t')
      ?.replaceAll("'", "\\\\'")
      ?.replaceAll('"', '\\\\"')
      ?.replaceAll('\\$', '\\\\\\$')
  }

  void visitConstantExpression(ConstantExpression expression, boolean unwrapQuotes = false, Boolean escapeChars = null) {
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 25]
```diff
@@ -0,0 +1,1241 @@
...
+      visitAnnotationNode (expression.value as AnnotationNode)
+      return
+    }
+    if (expression.value instanceof String && !unwrapQuotes) {
+      // string reverse escaping is very naive
+      def escaped = ((String)expression.value).replaceAll('\n', '\\\\n').replaceAll("'", "\\\\'")
+      print "'$escaped'"
+    } else {
+      print expression.value
+    }
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:890]:
```suggestion
    if ((expression.value instanceof String || expression.value instanceof Character) && !unwrapQuotes) {
      //noinspection GroovyPointlessBoolean
      String value = escapeChars == false ? expression.value : escapeCharacters(expression.value as String)
      print "'$value'"
    } else {
      print escapeChars == true ? escapeCharacters(expression.value as String) : expression.value
    }
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy [hunk 26]
```diff
@@ -0,0 +1,1241 @@
...
+
+  @Override
+  void visitGStringExpression(GStringExpression expression) {
+    // original groovy version loses enclosing ${}
+    // so we add it for everything, as the ${} style is always correct
+    def text = expression.text
+    expression.values.each {
+      text = text.replace("\$${it.text}", "\${${it.text}}")
+    }
+    print '"' + text + '"'
```
Comment (Vampire) [inline spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:933]:
```suggestion
    print '"'
    ((expression.strings as Collection<Expression>) + expression.values)
      .sort { left, right ->
        left.lineNumber <=> right.lineNumber
          ?: left.columnNumber <=> right.columnNumber
      }
      .each {
        if (it in expression.values) {
          if (it instanceof ClosureExpression) {
            print '$'
            it.visit(this)
          } else {
            print '${'
            if (it instanceof VariableExpression) {
              visitVariableExpression(it, false)
            } else {
              it.visit(this)
            }
            print '}'
          }
        } else if (it instanceof ConstantExpression) {
          visitConstantExpression(it, true, true)
        } else {
          it.visit(this)
        }
      }
    print '"'
```

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy (hunk 21 above)
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:428]: why the explicit `!= null` instead of simple groovy truth?

Comment (Vampire) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy:428]: Because it is a different case.
`ArrayList<>` for example.
With `if (generics)` it is `ArrayList` as the `generics` array is empty and thus Groovy falsy.
Wiht `if (generics != null)` it is rendered correctly as `ArrayList<>`.

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 27]
```diff
@@ -0,0 +1,630 @@
...
+    def y = (long)c
+    assert c == d
+  }
+
+  void gstrings() {
+    def a = "simple"
+    def b = "normal string \\$a ..."
+    def c = "gstring $a ..."
+    def d = "gstring with brackets ${a.size()} ..."
+    def e = "gstring with closure ${-> a} ... "
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:347]: You should also add cases where the closure has parameters

File: spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy [hunk 28]
```diff
@@ -0,0 +1,630 @@
...
+ with closure ${-> a}
+ with escaped closure \\${-> a}
+ ... """
+    def j = """multi line gstring
+ with simple value $a
+ with simple escaped value \\$a
+ with brackets ${a.size()}
+ with escaped brackets \\${a.size()}
+ with closure ${-> a}
+ with escaped closure ${-> a}
```
Comment (Vampire) [inline spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy:368]: from the text I guess this should be escaped?

File: spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy (hunk 14 above)
[the remark this answers is not in the document]
Comment (leonard84) [reply spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceConverter.groovy]: For the time being I would keep the current variant

**Review by Vampire — APPROVED** [review]
